jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetLambdaHandlerResolver.kt

```diff
@@ -0,0 +1,106 @@
+import com.jetbrains.rider.model.RdMethodInfo
+import com.jetbrains.rider.model.backendPsiHelperModel
+import com.jetbrains.rider.model.publishableProjectsModel
+import com.jetbrains.rider.projectView.ProjectModelViewHost
+import com.jetbrains.rider.projectView.solution
+import com.jetbrains.rider.run.configurations.method.getProjectModeId
+import com.jetbrains.rider.util.idea.application
+import software.aws.toolkits.jetbrains.services.lambda.LambdaHandlerResolver
+import software.aws.toolkits.jetbrains.services.lambda.dotnet.element.RiderLambdaHandlerFakePsiElement
+import java.time.Duration
+
+class DotNetLambdaHandlerResolver : LambdaHandlerResolver {
+
+    override fun version(): Int = 1
+
+    override fun findPsiElements(
+        project: Project,
+        handler: String,
+        searchScope: GlobalSearchScope
+    ): Array<NavigatablePsiElement> {
+
+        val handlerParts = handler.split("::")
+        if (handlerParts.size < 3) return emptyArray()
+
+        val assemblyName = handlerParts[0]
+        val type = handlerParts[1]
+        val methodName = handlerParts[2]
+
+        val projectModelViewHost = ProjectModelViewHost.getInstance(project)
+        val publishableProjects = project.solution.publishableProjectsModel.publishableProjects.values.toList()
+        val projectToProcess = publishableProjects.find { it.projectName == assemblyName } ?: return emptyArray()
+
+        val methodForHandler: RdMethodInfo? = project.solution.backendPsiHelperModel.findPublicMethod
+                .sync(MethodExistingRequest(
+                        className = type,
+                        methodName = methodName,
+                        targetFramework = "",
+                        projectId = projectModelViewHost.getProjectModeId(projectToProcess.projectFilePath)
+                ), RpcTimeouts.default)
+
+        val fieldId = methodForHandler?.fileId ?: return emptyArray()
+
+        return arrayOf(RiderLambdaHandlerFakePsiElement(project, methodName, fieldId))
+    }
+
+    override fun determineHandler(element: PsiElement): String? = null
+
+    override fun determineHandlers(element: PsiElement, file: VirtualFile): Set<String> = emptySet()
+
+    override fun isHandlerValid(project: Project, handler: String): Boolean {
+        val handlerParts = handler.split("::")
+        if (handlerParts.size < 3) return false
+
+        val assemblyName = handlerParts[0]
+        val type = handlerParts[1]
+        val methodName = handlerParts[2]
+
+        var isMethodExists = false
+
+        if (application.isDispatchThread) {
+            isMethodExists = isMethodExists(project, assemblyName, type, methodName)
+        } else {
+            var isCompleted = false
+            application.invokeLater {
+                isMethodExists = isMethodExists(project, assemblyName, type, methodName)
+                isCompleted = true
+            }
+
+            SpinWait.spinUntil(Lifetime.Eternal, Duration.ofMillis(1000)) {
```

Comment:
Am I correct in assuming that spinUnit spins until the lambda returns true?

Reply:
Correct. This implementation will return "false" in after a duration and will not produce an exception. I've increased duration to 2 second.